### ISSUE TITLE
Failing to bind buffer on Metal

### ISSUE THREAD

**Issue by DagAgren** [gfx-rs/wgpu#312]
When running the following code, the first time through loop, buffer 2 will not be bound to anything. On the second run through the loop, it does get set correctly:

	wgpu_render_pass_set_pipeline(renderpass, self->pipeline);
	wgpu_render_pass_set_bind_group(renderpass, 0, self->uniformBindGroup, NULL, 0);
	wgpu_render_pass_set_bind_group(renderpass, 2, self->vertexBuffer.bindGroup, NULL, 0);

	int startVertex = 2;
	for(int i = 0; i < numberOfTrails; i++) {
		int numberOfPoints = ((int *)self->pointCounts->data)[i];
		wgpu_render_pass_set_bind_group(renderpass, 1, self->instanceBuffer.bindGroup, (WGPUBufferAddress[]){ i * sizeof(TrailInstance) }, 1);

		wgpu_render_pass_draw(renderpass, 2 * numberOfPoints, 2, startVertex, 0);

		startVertex += 2 * (numberOfPoints + 2);
	}

Buffer 1 is re-bound each run through the loop, but the problem is with buffer 2, which is set once at the start. The Metal debugger captures this sequence of calls when the loop runs twice:

<img width="577" alt="Screenshot 2019-08-29 at 19 34 59" src="https://user-images.githubusercontent.com/11536828/63961032-7ba8a380-ca98-11e9-923b-30c3c52b0f0e.png">

Note how buffer 2 is not bound at all until after the first draw call.

Only the vertex shader uses buffer 2, so it only needs to be set for the vertex shader.

The pipeline layout is set up as follows:

	WGPUBindGroupLayoutId uniformlayout = wgpu_device_create_bind_group_layout(device, &(WGPUBindGroupLayoutDescriptor){
		.bindings = (WGPUBindGroupLayoutBinding[]) {
			{ .binding = 0, .visibility = WGPUShaderStage_VERTEX | WGPUShaderStage_FRAGMENT, .ty = WGPUBindingType_UniformBuffer },
		},
		.bindings_length = 1,
	});

	self->uniformBindGroup = wgpu_device_create_bind_group(device, &(WGPUBindGroupDescriptor) {
		.layout = uniformlayout,
		.bindings = (WGPUBindGroupBinding[]) {
			{ .binding = 0, .resource = { WGPUBindingResource_Buffer, .buffer = { { .buffer = self->uniformBuffer, .offset = 0, .size = uniformsize } } } },
		},
		.bindings_length = 1,
	});

	WGPUBindGroupLayoutId instancelayout = wgpu_device_create_bind_group_layout(device, &(WGPUBindGroupLayoutDescriptor){
		.bindings = (WGPUBindGroupLayoutBinding[]) {
			{ .binding = 0, .visibility = WGPUShaderStage_VERTEX | WGPUShaderStage_FRAGMENT, .ty = WGPUBindingType_UniformBufferDynamic },
		},
		.bindings_length = 1,
	});

	InitialiseBufferAndBindGroup(&self->instanceBuffer, instancelayout);

	WGPUBindGroupLayoutId vertexlayout = wgpu_device_create_bind_group_layout(device, &(WGPUBindGroupLayoutDescriptor){
		.bindings = (WGPUBindGroupLayoutBinding[]) {
			{ .binding = 0, .visibility = WGPUShaderStage_VERTEX, .ty = WGPUBindingType_StorageBuffer },
		},
		.bindings_length = 1,
	});

	InitialiseBufferAndBindGroup(&self->vertexBuffer, vertexlayout);

	WGPUPipelineLayoutId pipelinelayout = wgpu_device_create_pipeline_layout(device, &(WGPUPipelineLayoutDescriptor) {
		.bind_group_layouts = (WGPUBindGroupLayoutId[]) { uniformlayout, instancelayout, vertexlayout },
		.bind_group_layouts_length = 3,
	});

**Comment by DagAgren** [comment]
Inserting this immediately after the `wgpu_render_pass_set_bind_group(renderpass, 2,...` call fixes things:

	wgpu_render_pass_set_bind_group(renderpass, 1, self->instanceBuffer.bindGroup, (WGPUBufferAddress[]){ 1 * sizeof(TrailInstance) }, 1);

It does NOT fix it if the offset is 0, though.

**Comment by kvark** [comment]
I think I found the problem (or, more precisely, a few problems). Fix is coming, in the meantime here is the description of what happens.
This is the bind group [follow-up binding check](https://github.com/gfx-rs/wgpu/blob/b58c96e4a66943e09e9a9aa0ee8345c7bc878d38/wgpu-native/src/command/bind.rs#L181):
```rust
                        let end = if was_compatible {
                            trace!("\t\tgenerating follow-up sequence");
                            MAX_BIND_GROUPS
                        } else {
                            index + 1
                        };
```
What this is meant to do is: if this slot turned from being incompatible to being compatible, do the binding of all the higher-up bind groups ("follow-up bindings"), since they were blocked due to the pipeline compatibility rules of Vulkan and gfx-rs (TL;DR: higher descriptor sets depend on lower to be compatible).

But the code does the opposite of that: it returns the follow-up only when the bind group was *already* compatible. So when called again, this `was_compatible` became `true`, and we got the bindings in place.

The good news here is not only we are going to have this fixed, but also we were doing quite a bit of unnecessary binding before. So the fix will make us run faster ;)
